src/services/uniswap.js

```diff
@@ -0,0 +1,72 @@
+import Ethereum from '../services/eth';
+const uni = require('@uniswap/sdk')
+const ethers = require('ethers')
+const proxyArtifact = require('../static/uniswap_v2_router_abi.json')
+const debug = require('debug')('router')
+
+const GAS_LIMIT = 1200000  //120000000
+
+export default class Uniswap {
+  constructor (network = 'kovan') {
+    // network defaults to KOVAN
+    const providerUrl = process.env.ETHEREUM_RPC_URL
+    this.network = process.env.UNISWAP_NETWORK
+    this.chainID = uni.ChainId.KOVAN
+    this.provider = new ethers.providers.JsonRpcProvider(providerUrl)
+    this.exchangeProxy = process.env.UNISWAP_PROXY
+
+    if (network === 'kovan') {
+      this.chainID = uni.ChainId.KOVAN
+    } else if (network === 'mainnet') {
+      this.chainID = uni.ChainId.MAINNET
+    } else {
+      throw Error(`Invalid network ${network}`)
+    }
+  }
+
+  async fetch_route(tokenIn, tokenOut){
+      var route, pair, pairOne, pairTwo
+      var tIn = await uni.Fetcher.fetchTokenData(this.chainID, tokenIn)
+      var tOut = await uni.Fetcher.fetchTokenData(this.chainID, tokenOut)
+
+      try {
+      pair = await uni.Fetcher.fetchPairData(tIn, tOut)
+      route = new uni.Route([pair], tIn, tOut)
+      }
+      catch(err){
+      console.log('Trying alternative/indirect route.')
+      pairOne = await uni.Fetcher.fetchPairData(tIn, uni.WETH[this.chainID])
+      pairTwo = await uni.Fetcher.fetchPairData(tOut, uni.WETH[this.chainID])
+      route = new uni.Route([pairOne, pairTwo], tIn, tOut)
+      }
+      return route
+  }
+
+  async swapExactIn (wallet, route, tokenAddress, amountIn, gasPrice) {
+    const tIn = await uni.Fetcher.fetchTokenData(this.chainID, tokenAddress)
+    const tokenAmountIn = new uni.TokenAmount(tIn, amountIn)
+    const result = uni.Router.swapCallParameters(
+          uni.Trade.exactIn(route, tokenAmountIn),
+          { ttl: 50, recipient: wallet.address, allowedSlippage: new uni.Percent('1', '100') }
+        )
+
+    const erc = new Ethereum(this.network)
+    const approval = await erc.approveERC20(wallet, this.exchangeProxy, tokenAddress, ethers.utils.parseUnits(tokenAmountIn.toExact(), tIn.decimals), gasPrice)
+    let confirmApproval = await approval.wait()  // Wait to ensure that approval transaction is mined
```

Comment:
Also, is the approach of approving token amount just before swapping okay?

Reply:
The client automatically approves any unapproved assets when it starts operation, so we don't need to handle approval here. 

Instead, I would just add `uniswap/approve` and `uniswap/allowances` routes exactly like how I added them for Balancer in this PR: https://github.com/CoinAlpha/gateway-api/pull/10. Those functions are identical to what's in `eth/approve` and `eth/allowances`, except that they inject the `spender` address. This would allow the client to just call `[connector]/approve` and `[connector]/allowances`.